### PR TITLE
Core/Unit: Correct speed calculus when affected by SPELL_AURA_MOD_MINIMUM_SPEED

### LINKED ISSUE #4169: [Spell] Wings of the Damned
Hello,

There seems to be a slight bug, with handling of spell auras 305 (SPELL_AURA_MOD_MINIMUM_SPEED)

Right now, the code states:

``` c++
if (float minSpeedMod = (float)GetMaxPositiveAuraModifier(SPELL_AURA_MOD_MINIMUM_SPEED))
{
        float min_speed = minSpeedMod / 100.0f;
        if (speed < min_speed)
        speed = min_speed;
}
```

The point is, the value of the "Minimum speed" is taken as an absolute value, instead of as value relative to the unit's base speed rate.

ie. If we have two units,
unit A with base speed rate 1.0,
unit B with base speed rate 0.6, 

And we apply a -70% slow to both of them,

unit A will be slowed only by 50% ( down to 0.5 ), which is correct, but 
unit B will be slowed down to 0.5, instead of 50% of 0.6 ( 50% of base speed rate ).

How to quickly test in practice:
Place unit A and unit B, apply a slow debuff on both of them ( ie. Cone of Cold ), observe how they both run at same speed.

Note: I'm not 100% about correct expected behaviour, couldn't verify for or against it.

Thank you.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/65665-spell-wings-of-the-damned?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by sirikfoll**
**Changes proposed:**
New aproach is from the patch in https://github.com/TrinityCore/TrinityCore/issues/4169#issuecomment-2989390

**Target branch(es):** 3.3.5/master

**Issues addressed:** Closes #4169

**Tests performed:** Built and tested during the Lich King fight, with the Valkyr Shadowguard

**Known issues and TODO list:**

### CODE REVIEW

**Comment by ccrs** [comment]
check the interaction with the shaman spirit wolf spell and its related talent

**Comment by sirikfoll** [comment]
@ccrs  it definitely broke that :/

I know that set the Valkyr speed directly in the database may be considered a hack, but that, along with the second edit in unit.cpp, would solve the problem without break anything.
Any comments?

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -9166,7 +9166,7 @@ void Unit::UpdateSpeed(UnitMoveType mtype)
 
     if (float minSpeedMod = (float)GetMaxPositiveAuraModifier(SPELL_AURA_MOD_MINIMUM_SPEED))
     {
-        float min_speed = minSpeedMod / 100.0f;
+        float min_speed = (minSpeedMod / 100.0f) * GetSpeedRate(MOVE_RUN);
```
Comment (P-Kito) [inline src/server/game/Entities/Unit/Unit.cpp]: I suggest:
```cpp
CalculatePct(GetSpeedRate(MOVE_RUN), minSpeedMod);
```

**Comment by Aokromes** [comment]
@sirikfoll this can be merged?

**Comment by Shauren** [comment]
This is definitely incorrect in its current form. It breaks in case when movement speed slow aura is applied before SPELL_AURA_MOD_MINIMUM_SPEED. You would be treating the slowed speed as "normal"

**Comment by P-Kito** [comment]
Maybe introduce a new function that subtracts the negative speed mods and positive speed mods of a speedvalue. Then there would only be the DB value with scripted changes left. Just a random idea...

**Comment by Keader** [comment]
Any feedback here guys?

**Comment by Noryad** [comment]
Related to Valkyr Shadowguard speed is ok.

**Comment by Shauren** [comment]
Player pets need to ignore template base speed - same thing as with resistances applies here

**Comment by Noryad** [comment]
like 949b38b?

**Comment by sirikfoll** [comment]
Is it correct now?

File: src/server/game/Entities/Unit/Unit.cpp [hunk 2]
```diff
@@ -9268,7 +9268,11 @@ void Unit::UpdateSpeed(UnitMoveType mtype)
 
     if (float minSpeedMod = (float)GetMaxPositiveAuraModifier(SPELL_AURA_MOD_MINIMUM_SPEED))
     {
-        float min_speed = minSpeedMod / 100.0f;
+        float baseMinSpeed = 1.0f;
+        if (!(GetOwnerGUID().IsPlayer() || IsHunterPet()) && GetTypeId() == TYPEID_UNIT)
```
Comment (Shauren) [inline src/server/game/Entities/Unit/Unit.cpp]: please, dont try putting the negation outside, it makes the code hard to decipher

**Comment by Shauren** [comment]
Is good now.

**Pull request merged**